File: app/src/main/java/com/sdex/activityrunner/glide/ApplicationIconModelLoader.java
package com.sdex.activityrunner.glide;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.bumptech.glide.Priority;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.Options;
import com.bumptech.glide.load.data.DataFetcher;
import com.bumptech.glide.load.model.ModelLoader;
import com.bumptech.glide.signature.ObjectKey;
import com.sdex.activityrunner.db.application.ApplicationModel;

class ApplicationIconModelLoader implements ModelLoader<ApplicationModel, ApplicationModel> {

  @Nullable
  @Override
  public LoadData<ApplicationModel> buildLoadData(@NonNull final ApplicationModel applicationInfo,
                                                  int width, int height, @NonNull Options options) {
    return new LoadData<>(new ObjectKey(applicationInfo), new DataFetcher<ApplicationModel>() {
      @Override
      public void loadData(@NonNull Priority priority,
                           @NonNull DataCallback<? super ApplicationModel> callback) {
        callback.onDataReady(applicationInfo);
      }

      @Override
      public void cleanup() {

      }

      @Override
      public void cancel() {

      }

      @NonNull
      @Override
      public Class<ApplicationModel> getDataClass() {
        return ApplicationModel.class;
      }

      @NonNull
      @Override
      public DataSource getDataSource() {
        return DataSource.LOCAL;
      }
    });
  }

  @Override
  public boolean handles(@NonNull ApplicationModel applicationInfo) {
    return true;
  }
}

File: app/src/main/java/com/sdex/activityrunner/glide/ApplicationIconModelLoader.kt
package com.sdex.activityrunner.glide

import com.bumptech.glide.Priority
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.Options
import com.bumptech.glide.load.data.DataFetcher
import com.bumptech.glide.load.model.ModelLoader
import com.bumptech.glide.signature.ObjectKey
import com.sdex.activityrunner.db.cache.ApplicationModel

internal class ApplicationIconModelLoader :
    ModelLoader<ApplicationModel, ApplicationModel> {

    override fun buildLoadData(
        applicationInfo: ApplicationModel,
        width: Int, height: Int, options: Options
    ): ModelLoader.LoadData<ApplicationModel> {
        return ModelLoader.LoadData(
            ObjectKey(applicationInfo),
            object : DataFetcher<ApplicationModel> {
                override fun loadData(
                    priority: Priority,
                    callback: DataFetcher.DataCallback<in ApplicationModel>
                ) {
                    callback.onDataReady(applicationInfo)
                }

                override fun cleanup() {

                }

                override fun cancel() {

                }

                override fun getDataClass(): Class<ApplicationModel> {
                    return ApplicationModel::class.java
                }

                override fun getDataSource(): DataSource {
                    return DataSource.LOCAL
                }
            })
    }

    override fun handles(applicationInfo: ApplicationModel) = true
}
